components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core;
+
+import org.apache.axiom.om.OMElement;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.core.util.KeyStoreManager;
+import org.wso2.carbon.identity.core.model.IdentityKeyStoreMapping;
+import org.wso2.carbon.identity.core.util.IdentityConfigParser;
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverConstants;
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverConstants.InboundProtocol;
+import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverException;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.buildCustomKeyStoreName;
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.buildTenantKeyStoreName;
+import static org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverUtil.getQNameWithIdentityNameSpace;
+
+
+/**
+ * KeyStore manager for inbound authentication protocols.
+ * Retrieve keystores, private keys, public keys and public certificates.
+ */
+public class IdentityKeyStoreResolver {
+
+    private static IdentityKeyStoreResolver instance = null;
+
+    private static ConcurrentHashMap<InboundProtocol, IdentityKeyStoreMapping>
+            keyStoreMappings = new ConcurrentHashMap<>();
+
+    // Hashmaps to store retrieved private keys and certificates.
+    // This will reduce the time required to read configs and load data from keystores everytime.
+    private static Map<String, Key> privateKeys = new ConcurrentHashMap<>();
+    private static Map<String, Certificate> publicCerts = new ConcurrentHashMap<>();
+
+    private static Log log = LogFactory.getLog(IdentityKeyStoreResolver.class);
+
+    private IdentityKeyStoreResolver() {
+
+        parseIdentityKeyStoreMappingConfigs();
+    }
+
+    public static IdentityKeyStoreResolver getInstance() {
+
+        if (instance == null) {
+            instance = new IdentityKeyStoreResolver();
+        }
+        return instance;
+    }
+
+    /**
+     * Return Primary or tenant keystore according to given tenant domain.
+     *
+     * @param tenantDomain  Tenant domain
+     * @return Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public KeyStore getKeyStore(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
```

Comment:
What happens if tenantDomain becomes null ? (it is possible ?) 
if (tenantDomain == null || tenantDomain.isEmpty()) {
            throw new IllegalArgumentException("Tenant domain must not be null or empty");
        }

Reply:
Fixed https://github.com/wso2/carbon-identity-framework/pull/5794/commits/2f5713e26935f481b35b925fe8f81f2fe4e2f585